### PR TITLE
add health check

### PR DESCRIPTION
**Pull request by tonybase**
#2479

### CODE REVIEW

**Comment by shenqidebaozi** [comment]
动态控制健康检查结果是先不实现吗？还是通过其他方式实现

**Comment by shenqidebaozi** [comment]
代码很清真，看起来只需要修改为通过 grpc/http 借口将健康结果暴露即可，另外在服务生命周期是不是也要控制健康检查结果？

File: health/health.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+// CheckerFunc is an adapter type to allow the use of ordinary functions as
+// health checks.
+type CheckerFunc func() error
+
+// CheckHealth calls f().
+func (f CheckerFunc) CheckHealth() error {
+	return f()
+}
```
Comment (jcyamacho) [inline health/health.go:39]: 🤔 maybe `ctx` param is mising in the adapter

File: health/health.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+		if err := c.CheckHealth(ctx); err != nil {
+			res.Details[n] = Detail{Status: StatusDown, Error: err.Error()}
+		} else {
+			res.Details[n] = Detail{Status: StatusUp}
+		}
+	}
+	return res
+}
+
+func (h *Health) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```
Comment (jcyamacho) [inline health/health.go:82]: it's ok to not have an option for creating 2 different handlers (liveness and readiness) ??
or the idea is to create 2 different `Health` and add only one category to each one??

Comment (tonybase) [reply health/health.go:82]: I think there's only liveness, we only deal with application health, and components are unavailable and measured by metrics, logs.

**Pull request closed without merging**